### ISSUE TITLE
H.265 feedback and possible settings for --quick option

### ISSUE THREAD

**Issue by vickash** [lisamelton/video_transcoding#238]
After reading #229, I decided to test if H.265 would work for me since all but one of my Plex clients is capable of 8-bit H.265 and space savings would be significant.

@donmelton, in that thread you recommend 10-bit, which isn't an option for me, and using `---abr`. I wanted to test what would happen if I used your CRF controls for x264 anyway. It transfers over well as far as I can tell. All I did at first was change the encoder, and pass `--target 3000` in an attempt to halve my bitrate for 1080p Blu Ray rips.

As for speed, I eventually started using the following options, taking hints from #182 :
`-E rc-lookahead=15`
`-E ref=1`
`-E b-adapt=0`
`-E rd=2`
`-E fast-intra=1`
`-E early-skip=1`
Sub-pixel motion estimation is left at the "medium" preset, while everything else comes down to the "superfast" preset. Doing this almost doubles encoding speed. Maybe this would make a good starting point for adding a `--quick` option for H.265?

Lastly, I'm not 100% sure if `-E qpmax=34` is valid for H.265. The [docs](https://x265.readthedocs.io/en/default/cli.html) make no mention of that option, but x265 doesn't warn about it. The docs do contain `-E crf-max=34`.  I've been adding this for a while now just to be safe, but didn't do an A/B test to see if it mattered. Will test and let you know.

**Comment by donmelton** [comment]
@vickash Thanks for experimenting with this. I'll try to take a look at the options soon.

But x265 definitely supports the `qpmax` option. They added it over a year ago after I bugged the head of the project about it. :) Just execute `x265 --fullhelp` to see it.

**Comment by vickash** [comment]
Yes. I just found that thread and was about to correct myself. `qpmax` is in the PDF docs, but not the HTML which I was reading.

I also must have misread your rate control explanation and set `crf-max` to the same value as `qpmax`. You're using 25 on x264. Luckily, I knew something was wrong and 34 looked bad enough that I started using a value of 28 for both early on. I won't bother re-encoding those as they look good, but will fix going forward.

I wonder if `qpmax` might need to be lowered for x265 compared to x264. x265 should look better than x264 for a given QP value, but because we're trying to get lower output bitrates out of x265, it seems like the same `qpmax` value causes the encoder to be more aggressive and results in a higher variance of quality. Maybe better to tradeoff some bitrate to get it back.

**Comment by donmelton** [comment]
@vickash Actually, it's possible that I would raise both `crf-max` and `qpmax` when transcoding with x265. The reason is that this would allow for lower bitrate targets, which is what most people want from HEVC. And, technically, quality should be maintained since HEVC is _supposed_ to be higher quality when the same quantizer value is used. I am dubious about that though.

But, really, you should be using `--abr` or `--simple` with the x265 encoder. Those work much better than my special, or default ratecontrol system at the moment.

Also, x265 will likely still be slower than shit even if we come up with a `--quick` implementation that maintains quality. :) For HEVC output, I think the new available hardware-based encoders are a much better choice for most users. For example, on my MacBook Pro, I get both great quality and faster-than-real-time transcoding speeds using the `vt_h265` encoder.

**Comment by donmelton** [comment]
@vickash BTW, I have not forgotten about this. It's just mind-numbingly tedious to test x265 on a slow machine. Which is what I have.

I just wanted you to know. :)

**Comment by hazcod** [comment]
Out of curiosity, what was the consensus about hevc 10bit? Is constant bitrate used?

**Comment by donmelton** [comment]
@hazcod Sorry I took so long to reply. I don't use true CBR (constant bitrate) for any of my ratecontrol systems. They're all bitrate-targeted but will vary somewhat. And it doesn't matter whether that's 8-bit, 10-bit or 12-bit.

**Comment by TehMaster** [comment]
Hi Don - Just wondering if anything new has come of this, and what the general consensus is for H.265 settings? Is it enough to simply use the --avbr flag, or are more flags needed such as the crf-max, qpmax, etc?

**Comment by donmelton** [comment]
@TehMaster The general consensus is to use a hardware-based encoder for HEVC output. IMO, the `x265` encoder is still way too slow to be viable for home transcoding.

Please check out my new project: https://github.com/donmelton/other_video_transcoding

...It's designed to leverage hardware encoders much easier.

Also, the `--avbr` option only works for H.264 output.

**Comment by TehMaster** [comment]
@donmelton Thanks Don; appreciate the info. I guess i'll simply wait until smarter minds get this all sorted out and wrapped in an easy to use command or two... :)
